### PR TITLE
feat: get_extra_metrics

### PR DESCRIPTION
**Pull request by garrett361**
Adds a `ModelProtocol.get_extra_metrics` method for more flexible custom metric reporting, as discussed in #1576 

Probably this should be an abstract method, but I was wary of making this a breaking change for users who inherit this commit. 

The current signature is `get_extra_metrics(self, parallel_dims: ParallelDims) -> None | dict`. I also considered adding some subset of `JobConfig`, `TrainSpec`, and `pp_has_{first,last}_stage`; not sure what else might be useful.

Tested via running the debugmodel with print statements. 

CC @rakkit @wwwjn

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -518,6 +518,10 @@ def train_step(
             "n_tokens_seen": global_ntokens_seen,
             "lr": lr,
         }
+        for mp in self.model_parts:
+            if mp_metrics := mp.get_extra_metrics(self.parallel_dims):
+                extra_metrics.update(mp_metrics)
```
Comment (fegin) [inline torchtitan/train.py]: This approach seems quite fragile. What happens if two modules return metrics with the same keys? This could lead to silent overwriting of data.

File: torchtitan/protocols/model.py [hunk 2]
```diff
@@ -53,3 +54,6 @@ def init_weights(self, buffer_device: torch.device | None = None) -> None:
             buffer_device: Optional device to place buffers on during initialization.
         """
         pass
+
+    def get_extra_metrics(self, parallel_dims: ParallelDims) -> None | dict:
```
Comment (fegin) [inline torchtitan/protocols/model.py]: The return value should be `None | dict[str, Any]`

**Review by fegin — COMMENTED** [review]
The idea looks good to me but I have concerns about how we are going to deal with pipeline parallelism. It can lead to confusing metrics report.

File: torchtitan/train.py (hunk 1 above)
Comment (garrett361) [reply torchtitan/train.py]: Yes, I have the same concern, but was not sure of how to handle this case, apart from `warning`.  Was hoping to have a discussion about that here

File: torchtitan/protocols/model.py (hunk 2 above)
Comment (garrett361) [reply torchtitan/protocols/model.py]: Done. 

I meant to revisit this; I wasn't sure what kind of restrictions `wandb` had on the values, but `Any` is fine for now.

File: torchtitan/train.py (hunk 1 above)
Comment (garrett361) [reply torchtitan/train.py]: This is why I was considering adding the other objects to the `get_extra_metrics` signature above, so you could adjust reporting based on PP scheduler, etc, also.

The logic is going to get pretty complex if a user's `get_extra_metrics` needs to handle all possible sharding cases, but that also seems unavoidable, and at least some API like this could make the path to custom metrics less painful.

File: torchtitan/protocols/model.py [hunk 3]
```diff
@@ -53,3 +54,6 @@ def init_weights(self, buffer_device: torch.device | None = None) -> None:
             buffer_device: Optional device to place buffers on during initialization.
         """
         pass
+
+    def get_extra_metrics(self, parallel_dims: ParallelDims) -> None | dict[str, Any]:
```
Comment (tianyu-l) [inline torchtitan/protocols/model.py]: Changing protocols should be done carefuly:
- In general `ModelProtocol` shouldn't be parallelism-aware, so maybe let's not pass in `ParallelDims` explicitly. We can make it extensible by doing `*args` and `**kwargs`
- `get_extra_metrics` may not be a good name because here there are no basic metrics, so meaning of "extra" is not clear.

File: torchtitan/train.py [hunk 4]
```diff
@@ -518,6 +518,10 @@ def train_step(
             "n_tokens_seen": global_ntokens_seen,
             "lr": lr,
         }
+        for mp in self.model_parts:
```
Comment (tianyu-l) [inline torchtitan/train.py]: > The idea looks good to me but I have concerns about how we are going to deal with pipeline parallelism.

@fegin 
maybe PP related logic can go here -- we can update the keys in the for-loop.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Thanks for the proposal. Had some comments.

File: torchtitan/protocols/model.py (hunk 3 above)
Comment (garrett361) [reply torchtitan/protocols/model.py]: Yeah, the naming was based on the `extra_metrics` arg of `MetricsProcessor.log`, but then this needlessly couples names between the protocol and the processor class

Maybe this is better implemented as a new method on `MetricsProcessor`? And users can subclass and override that class to implement custom metrics logic. 

```py
class MetricsProcessor:
    [...]
    def get_extra_metrics(self, model_parts: list[nn.Module], *args, **kwargs) -> None  | dict[str, Any]:
        return
``` 
and 
```py
# train.py

extra_metrics = {
    "n_tokens_seen": global_ntokens_seen,
    "lr": lr,
}
if mp_extra_metrics := self.metrics_processor.get_extra_metrics(self.model_parts):
    extra_metrics.update(mp_extra_metrics)

self.metrics_processor.log(
    self.step,
    global_avg_loss,
    global_max_loss,
    grad_norm.item(),
    extra_metrics=extra_metrics,
)
```
Don't love it either, but implementing via `MetricsProcessor` feels more natural to me than via `ModelProtocol`. What do you think?

`MetricsProcessor` already holds a ref to `ParallelDims`, `JobConfig` and more, which is nice. 

CC @fegin

Comment (tianyu-l) [reply torchtitan/protocols/model.py]: I also thought more and had the same idea of putting it outside `ModelProtocol`. Logging is really not the responsibility of models anyway.

Putting it into `metrics_processor` sounds reasonable to me. Also I think if that's the case, we don't have to modify `train.py` -- we can just call `get_extra_metrics` inside `self.metrics_processor.log()` as it doesn't mutate the model anyway.

We can reason more about naming. Having multiple `extra_metrics` still sounds confusing.

Comment (garrett361) [reply torchtitan/protocols/model.py]: > We can reason more about naming. Having multiple extra_metrics still sounds confusing.

Agreed. 

> we don't have to modify train.py

Good point, also agreed.  **EDIT**: though were need to get the metrics processor a reference to model parts somehow still.

Comment (garrett361) [reply torchtitan/protocols/model.py]: `MetricsProcessor.get_model_parts_metrics`?

Comment (fegin) [reply torchtitan/protocols/model.py]: Putting this to `MetricsProcessor` looks more nature. As for the naming, I was originally thinking `get_custom_metrics()` but realized that it is too general and we won't know what arguments the trainer should pass to this function to get the custom metrics. In the end, users may need to change the trainer if this function allows arbitrary metrics.

If this function can get the metrics exclusively from model parts, `get_model_parts_metrics` looks fine, but I would suggest a slightly lengthy one, `get_model_parts_custom_metrics`, as there are already metrics derived from model parts.

Comment (tianyu-l) [reply torchtitan/protocols/model.py]: I'd vote for `model` instead of `model_parts`. `model_parts` are PP-related details which shouldn't go into generic function names (I understand they need to go into args).
Although,  I'm not sure about `get_custom_model_metrics` or `get_model_custom_metrics`.

Comment (fegin) [reply torchtitan/protocols/model.py]: Both names look okay to me. I feel this falls into the English grammar discussion, which I'm pretty bad at this topic.

Comment (garrett361) [reply torchtitan/protocols/model.py]: I am fine with any of these names. Slight preference for `get_custom_model_metrics`. I also don't have a problem with `get_custom_metrics`.

I'm out sick at the moment, but will change this PR to implement custom metrics via `MetricsProcessor` when I'm back at work.

**Comment by garrett361** [comment]
Updated: now implemented via `MetricsProcessor.get_custom_model_metrics`. CC @fegin @tianyu-l @rakkit

File: torchtitan/train.py [hunk 5]
```diff
@@ -518,6 +519,15 @@ def train_step(
             "n_tokens_seen": global_ntokens_seen,
             "lr": lr,
         }
+        if custom_model_metrics := self.metrics_processor.get_custom_model_metrics(
+            self.model_parts
+        ):
+            for k, v in custom_model_metrics.items():
+                if k in extra_metrics:
+                    warnings.warn(
```
Comment (fegin) [inline torchtitan/train.py]: Can we change this to `logger.warning` to align other use cases in TorchTitan?

Comment (garrett361) [reply torchtitan/train.py]: yeah, makes sense

Comment (garrett361) [reply torchtitan/train.py]: done @fegin

File: torchtitan/components/metrics.py [hunk 6]
```diff
@@ -454,6 +456,11 @@ def log_validation(self, loss: float, step: int):
         self.time_last_log = time.perf_counter()
         self.device_memory_monitor.reset_peak_stats()
 
+    def get_custom_model_metrics(
+        self, model: nn.Module | list[nn.Module], *args, **kwargs
+    ) -> dict[str, Any] | None:
+        return None
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: Please add a comment / docstring to explain what this placeholder function if for.

File: torchtitan/train.py [hunk 7]
```diff
@@ -518,6 +518,15 @@ def train_step(
...
         }
+        if custom_model_metrics := self.metrics_processor.get_custom_model_metrics(
+            self.model_parts
+        ):
+            for k, v in custom_model_metrics.items():
+                if k in extra_metrics:
+                    logger.warning(
+                        f"Overriding extra_metrics['{k}']={extra_metrics[k]} with value {v}"
+                    )
+                extra_metrics[k] = v
```
Comment (tianyu-l) [inline torchtitan/train.py]: I'd suggest let's not do this in `train.py`, as it's meaningless for users who didn't overwrite `get_custom_model_metrics` -- we can do it in `self.metrics_processor.log()`. You can make `model_parts` a variable of `MetricsProcessor`, as it already has `optimizer` and `lr_scheduler`.

One step further: since you anyways will define your own `MetricsProcessor`, it's not clear why do you need torchtitan to host `get_custom_model_metrics`. The only critical thing is to pass `model_parts` to `MetricsProcessor` which I'm OK with.

Comment (tianyu-l) [reply torchtitan/train.py]: cc @fegin in case you think otherwise

Comment (garrett361) [reply torchtitan/train.py]: I'd also be fine to change this PR to be super minimal and just introduce `model_parts` as a `MetricsProcessor` class attr. I'd actually prefer that, if everyone else is on board.

Comment (fegin) [reply torchtitan/train.py]: Sounds good to me.

**Comment by garrett361** [comment]
Alright, now this PR just adds a `model_parts` attr to `MetricsProcessor` and `train.py` populates the value. 

This is much better, imo: minimal and not prescriptive for how the user has to log additional metrics. Thanks for the discussion @fegin @tianyu-l

**Review by tianyu-l — APPROVED** [review]
LGTM, thank you!

**Comment by garrett361** [comment]
fixed a linting issue
